wagtail/wagtailcore/query.py

```diff
@@ -84,6 +84,105 @@ def not_ancestor_of(self, other, inclusive=False):
         """
         return self.exclude(self.ancestor_of_q(other, inclusive))
 
+    def first_common_ancestor(self, include_self=False, strict=False):
```

Comment:
I moved it from `PageQuerySet` to `TreeQuerySet` since this logic is generic enough to be used for any treebeard tree. This way it can be used for collection tree navigation.